### ISSUE TITLE
Decript problem if hash contains "+"

### ISSUE THREAD

**Issue by lukasjuhas** [o/crypt-php#1]
Hey there,

Thank you very much for your code, I appreciate that the project haven't been updated for the past 5 years but I'm wondering if you could help me out. 

Basically I discovered that whenever I  decrypt  and hasch conains "+", it fails, getting exception :

```
( ! ) Fatal error: Uncaught exception 'Exception' with message 'The given data appears tampered or corrupted '.....' on line 161
( ! ) Exception: The given data appears tampered or corrupted in '.....' on line 161
```

In my testing, I saved all the hashes that were generated, they were saved as a cookie. (I have changed prefix)

These worked:

```
wcl_#NhhMHpVj/IE7De7fvjIqmJFb4xAe3lC/#OEX5#91b8e01b8f1f26ff842d574b20a4d7a12c4933ec:"1"
wcl_#hTJNj7iEw9hOU9woA8FTcayXG8JhAnZD#2o4m#67003152e74faa6b818154d1cd8b0c7f1e857eaa:"1"
wcl_#zrmG9WoyGEIvVu68qFah5POebrAzXGpI#DhzF#2693f855a7bf4db1ea2fb44f0e2be7ff79adb0ee:"1"
wcl_#UzgSEAi7IUoSyqgt1Qm9JdVhuozahz9L#Tn3a#160c4530eeedff4659bf7cf87d9b2c4d0fbec2f5:"1"
wcl_#GHr6zWcwJTpZG3w0RRp5tz7wePIhvmsr#mteQ#3ce33e4516379f7f756012299af997a67481db71:"1"
wcl_#ypDTjze3EwJy94i5uJZAxXSBWUkuITYd#HqLY#a0eb817b83a46223f6a175a491a2798a331aeb24:"1"
```

These did not work:

```
wcl_#axJF6dU4+vR4x8rfNm/6JVAb1JAUulbp#Gr/j#f0713324dd4b13053d6abc13c4102dc76ff573a9:"1"
wcl_#35vj0QWPlyCOFH43t42Mda+jpABeu1e9#CEgA#6a011362e9b29a58fe11f218a6a023f5b62273d3:"1
```

Any ideas?

For now, I solved it with a quick hack, by repeating function if the hash contains "+" which is not very good solution.

Thanks! 

Lukas

**Comment by o** [comment]
Hello Lukas!

Which OS and PHP version you're using? 

I'm currently on OSX 10.11.3 - PHP 7.0.2 and tried to run (and update) decrypt examples with the your scenario. example: https://github.com/o/crypt-php/blob/master/examples/decryptScalar.php#L5 and works for me.

Can you try same decrypt examples in your box?

"+" is completely valid base64 string char. I suspect this is a platform dependent problem with base64 encoding / decoding or mcrypt. 
This is mine mcrypt.

```
$ mcrypt --version
Mcrypt v.0.9.9 (i386-apple-darwin15.0.0)
Linked against libmcrypt v.2.5.8
```

**Comment by lukasjuhas** [comment]
Hello!

Thank you for your prompt reply.

I'm running it on virtual box - laravel / homestead nginx server.

```
vagrant@homestead:~$ php -v
PHP 5.6.10-1+deb.sury.org~trusty+1 (cli)
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2015, by Zend Technologies
    with Xdebug v2.3.2, Copyright (c) 2002-2015, by Derick Rethans
    with blackfire v1.7.0, https://blackfire.io, by Blackfireio Inc.
```

```
vagrant@homestead:~$ mcrypt --version
Mcrypt v.2.6.8 (x86_64-pc-linux-gnu)
Linked against libmcrypt v.2.5.8
Copyright (C) 1998-2002 Nikos Mavroyanopoulos (nmav@gnutls.org)
```

I'm going to try your examples, I'll let you know how did that go.

Thanks!

Lukas

**Comment by lukasjuhas** [comment]
Hi there,

I'm having same problem with examples, same error. Any ideas? It's strange...

Thank you,

Lukas

**Comment by o** [comment]
Currently i haven't any idea about that, just setting up a new server with PHP 5.6 and i'll try examples.

**Comment by o** [comment]
Here is the results. Can you try to update your homestead installation with latest PHP?

```
polimat root ~/crypt-php/examples (master=) $ php -v
PHP 5.6.17-0+deb8u1 (cli) (built: Jan 13 2016 09:10:12) 
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2015, by Zend Technologies
polimat root ~/crypt-php/examples (master=) $ php decryptMixed.php 
object(ArrayObject)#2 (1) {
  ["storage":"ArrayObject":private]=>
  array(3) {
    [0]=>
    string(3) "foo"
    [1]=>
    string(3) "bar"
    [2]=>
    string(3) "baz"
  }
}
polimat root ~/crypt-php/examples (master=) $ uname -a
Linux polimat.org 4.1.5-x86_64-linode61 #7 SMP Mon Aug 24 13:46:31 EDT 2015 x86_64 GNU/Linux
```

**Comment by lukasjuhas** [comment]
Hello,

thank you,

so the problem is PHP compatibility. That might be a problem as I can't use PHP 7 on my current project. Do you have any ideas of workaround? 

Lukas

**Comment by o** [comment]
Can you try to update newest PHP 5.6 on your Vagrant? 

You can see online results from here https://3v4l.org/rT7Zm
Looks like it works, without any problem with this php versions `5.4.12 - 7.0.3`

**Comment by lukasjuhas** [comment]
Perfect, that seems to be working now. Updating to php 5.6.18 solved the problem.

```
vagrant@homestead:~$ php -v
PHP 5.6.18-1+deb.sury.org~trusty+1 (cli)
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2016, by Zend Technologies
    with Xdebug v2.3.2, Copyright (c) 2002-2015, by Derick Rethans
    with blackfire v1.7.0, https://blackfire.io, by Blackfireio Inc.
```

Thanks a lot for your help!

Lukas

**Comment by o** [comment]
Glad to see this helps. Cheers.

**Comment by lukasjuhas** [comment]
Hi there,

Unfortunately, the problem came back! Don't know why I didn't notice it yesterday. 

But.. I think I have found solution to my problem, I have created a fork and committed the change that helped me solve my problem.

[https://github.com/lukasjuhas/crypt-php/commit/058da1085f6954b304399e9360711363b2e5eff3](https://github.com/lukasjuhas/crypt-php/commit/058da1085f6954b304399e9360711363b2e5eff3)

Thanks!
